src/main/java/bio/terra/service/filedata/google/gcs/GcsPdao.java

```diff
@@ -674,19 +713,41 @@ public void blobAclUpdates(
     }
   }
 
+  /**
+   * Non-static version of {@link GcsPdao#getBlobFromGsPath(Storage, String, String)} to enable
+   * mocking in testing
+   */
+  public Blob getBlobFromGsPathNs(Storage storage, String gspath, String targetProjectId) {
+    return getBlobFromGsPath(storage, gspath, targetProjectId);
+  }
+
   public static Blob getBlobFromGsPath(Storage storage, String gspath, String targetProjectId) {
     BlobId locator = GcsUriUtils.parseBlobUri(gspath);
 
     // Provide the project of the destination of the file copy to pay if the
     // source bucket is requester pays.
-    Blob sourceBlob = storage.get(locator, Storage.BlobGetOption.userProject(targetProjectId));
+    BlobGetOption[] getOptions = new BlobGetOption[0];
+    if (targetProjectId != null) {
+      getOptions = new BlobGetOption[] {BlobGetOption.userProject(targetProjectId)};
+    }
+    Blob sourceBlob = storage.get(locator, getOptions);
     if (sourceBlob == null) {
       throw new PdaoSourceFileNotFoundException("Source file not found: '" + gspath + "'");
     }
 
     return sourceBlob;
   }
 
+  /**
+   * Extract the project to use to access a blob (using the userProject query param)
+   *
+   * @param gspath a gs:// path to a file with optional query parameters
+   * @return the value of the userProject query parameter or null
+   */
+  public static String getProjectIdFromGsPath(String gspath) {
+    return UriUtils.getValueFromQueryParameter(gspath, "userProject");
```

Comment:
Could this use a named constant?
```suggestion
    return UriUtils.getValueFromQueryParameter(gspath, GcsConstants.USER_PROJECT_QUERY_PARAM);
```

Reply:
Gah, great catch.  I thought I'd caught all of them